osu.Game.Tests/Visual/Gameplay/TestScenePlayerMaxDimensions.cs

```diff
@@ -63,18 +72,66 @@ public UpscaledLegacySkin(IStorageResourceProvider resources)
             }
 
-            public override Texture? GetTexture(string componentName, WrapMode wrapModeS, WrapMode wrapModeT)
+            public ISkin FindProvider(Func<ISkin, bool> lookupFunction) => this;
+            public IEnumerable<ISkin> AllSources => new[] { this };
+
+            protected override IResourceStore<TextureUpload> CreateTextureLoaderStore(IStorageResourceProvider resources, IResourceStore<byte[]> storage)
+                => new UpscaledTextureLoaderStore(base.CreateTextureLoaderStore(resources, storage));
+
+            private class UpscaledTextureLoaderStore : IResourceStore<TextureUpload>
             {
-                var texture = base.GetTexture(componentName, wrapModeS, wrapModeT);
+                private readonly IResourceStore<TextureUpload>? textureStore;
 
-                if (texture != null)
-                    texture.ScaleAdjust /= 8f;
+                public UpscaledTextureLoaderStore(IResourceStore<TextureUpload>? textureStore)
+                {
+                    this.textureStore = textureStore;
+                }
 
-                return texture;
-            }
+                public void Dispose()
+                {
+                    textureStore?.Dispose();
+                }
 
-            public ISkin FindProvider(Func<ISkin, bool> lookupFunction) => this;
-            public IEnumerable<ISkin> AllSources => new[] { this };
+                public TextureUpload Get(string name)
+                {
+                    var textureUpload = textureStore?.Get(name);
+
+                    // NRT not enabled on framework side classes (IResourceStore / TextureLoaderStore), welp.
+                    if (textureUpload == null)
+                        return null!;
+
+                    return upscale(textureUpload);
+                }
+
+                public async Task<TextureUpload> GetAsync(string name, CancellationToken cancellationToken = new CancellationToken())
+                {
+                    // NRT not enabled on framework side classes (IResourceStore / TextureLoaderStore), welp.
+                    if (textureStore == null)
+                        return null!;
+
+                    var textureUpload = await textureStore.GetAsync(name, cancellationToken).ConfigureAwait(false);
+
+                    if (textureUpload == null)
+                        return null!;
+
+                    return await Task.Run(() => upscale(textureUpload), cancellationToken).ConfigureAwait(false);
+                }
+
+                private TextureUpload upscale(TextureUpload textureUpload)
+                {
+                    var image = Image.LoadPixelData(textureUpload.Data.ToArray(), textureUpload.Width, textureUpload.Height);
+
+                    // The original texture upload will no longer be returned or used.
+                    textureUpload.Dispose();
+
+                    image.Mutate(i => i.Resize(new Size(textureUpload.Width, textureUpload.Height) * scale_factor));
```

Comment:
This doesn't match the title of the PR. This is not a crop operation, it's a resize still. Which is good, since cropping as I understand it would be _actively_ wrong, but anyhow, what's going on here? Can you explain _precisely_ what's causing the issue this is intending to fix?

Reply:
This is the test scene upscaling legacy skin elements to be tested with the skin's cropping mechanism, it's not the fix of this PR. See the change in `LegacySkinExtensions`.

Also, I stated already what this is supposed to fix in OP.

Reply:
OP says:

> The size is already larger than the specified maximum size for hit circles (128x128, which affects the glyph sprites). Cropping the texture from the centre fixes the issue.

I don't understand how this is supposed to fix what from the screenshot looks like an invalid hit circle number overlap. Can you elaborate on that part?

Also I have a feeling that this is a fix that's only correct to this specific use case and will break many others. I wouldn't be surprised if this is somehow specific to hitcircle numbers.

Reply:
It may seem like it, but the actual reason why it's "broken" is because we're downscaling any glyph in the hit circle's sprite text if it's larger than 128x128 (see [this](https://github.com/ppy/osu/blob/09c9baadec535bed2869d897764d340902868932/osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs#L142)). This PR changes the logic to use cropping instead of downscaling.

I can't really think of a case where this is worse than scaling. In fact, I was wondering why I picked scaling in the first place. In osu!(stable), a skin with elements that have extra space on all sides would still appear appropriately in gameplay, as opposed to elements that are upscaled. So cropping should be a better approach to be compatible with stable skins.

Reply:
It appears the change in spacing between hitcircle numbers is occurring due to the following sequence of code snippets:

<img width="1632" alt="CleanShot 2023-09-30 at 13 16 53@2x" src="https://github.com/ppy/osu/assets/22781491/d8247474-c0e5-4ab3-bc8d-cefbeb370060">
<img width="1610" alt="CleanShot 2023-09-30 at 13 16 30@2x" src="https://github.com/ppy/osu/assets/22781491/dd0ce106-a689-4287-97e9-e51e884fa55e">
<img width="464" alt="CleanShot 2023-09-30 at 13 14 32@2x" src="https://github.com/ppy/osu/assets/22781491/8f0bae16-a577-4cb6-bd84-4715998ec5a6">

TL;DR: `LegacySpriteText` uses `texture.ScaleAdjust` to adjust the kerning to be correct. The higher the downscale ratio, the lower the kerning is. Whether that's correct or not, I'm not sure. Enforcing all elements to `ScaleAdjust = 1` show abnormal space between glyphs in gameplay, but I don't want to look at this any further since we haven't been changing `ScaleAdjust` anywhere else besides in `LegacySkin.GetTexture` (and this PR will remove the usage in `WithMaximumSize`), and I don't want to derail this conversation over such behaviour.